crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+          , C.SomeHandle cfgHandle == C.frameHandle (sim_state ^. C.stateCrucibleFrame)
+          , Just Refl <- W4.testEquality
+              (fmapFC C.blockInputs cfgBlockMap)
+              (fmapFC C.blockInputs $ C.frameBlockMap $ sim_state ^. C.stateCrucibleFrame)
+          , elem true_frame_some_block_id loop_body_some_block_ids /= elem false_frame_some_block_id loop_body_some_block_ids -> do
+
+            (loop_condition, inside_loop_frame, _outside_loop_frame) <-
+              if elem true_frame_some_block_id loop_body_some_block_ids
+              then
+                return (branch_condition, true_frame, false_frame)
+              else do
+                not_branch_condition <- W4.notPred sym branch_condition
+                return (not_branch_condition, false_frame, true_frame)
+
+            case fixpoint_state of
+              BeforeFixpoint -> C.panic "SimpleLoopInvariant.simpleLoopInvariant:" ["BeforeFixpoint"]
+
+              ComputeFixpoint _fixpoint_record -> do
+                -- continue in the loop
+                ?logMessage $ "SimpleLoopInvariant: SymbolicBranchState: ComputeFixpoint"
+
+                loc <- W4.getCurrentProgramLoc sym
+                C.addAssumption bak $ C.BranchCondition loc (C.pausedLoc inside_loop_frame) loop_condition
+
+                C.ExecutionFeatureNewState <$>
+                  runReaderT
+                    (C.resumeFrame (C.forgetPostdomFrame inside_loop_frame) $ sim_state ^. (C.stateTree . C.actContext))
+                    sim_state
+
+              CheckFixpoint _fixpoint_record -> do
+                ?logMessage $ "SimpleLoopInvariant: SymbolicBranchState: CheckFixpoint"
+
+                return C.ExecutionFeatureNoChange
+
+      _ -> return C.ExecutionFeatureNoChange
+
+
+advanceFixpointState ::
+  forall sym bak ext p rtp blocks r args .
+  (C.IsSymBackend sym bak, C.HasLLVMAnn sym, ?memOpts :: C.MemOptions, ?logMessage :: String -> IO ()) =>
+  bak ->
+  C.GlobalVar C.Mem ->
+  (InvariantPhase -> [Some (W4.SymExpr sym)] -> MapF (W4.SymExpr sym) (InvariantEntry sym) -> IO (W4.Pred sym)) ->
+  C.BlockID blocks args ->
+  C.SimState p sym ext rtp (C.CrucibleLang blocks r) ('Just args) ->
+  FixpointState p sym ext rtp blocks ->
+  IORef (FixpointState p sym ext rtp blocks) ->
+  IO (C.ExecutionFeatureResult p sym ext rtp)
+
+advanceFixpointState bak mem_var loop_invariant block_id sim_state fixpoint_state fixpoint_state_ref = do
+  let ?ptrWidth = knownNat @64
+  let sym = C.backendGetSym bak
+  loc <- W4.getCurrentProgramLoc sym
+  case fixpoint_state of
+    BeforeFixpoint -> do
+      ?logMessage $ "SimpleLoopInvariant: RunningState: BeforeFixpoint -> ComputeFixpoint " ++ show block_id ++ " " ++ show (pretty (W4.plSourceLoc loc))
+      assumption_frame_identifier <- C.pushAssumptionFrame bak
+      let mem_impl = fromJust $ C.lookupGlobal mem_var (sim_state ^. C.stateGlobals)
```

Comment:
Any `fromJust` always scares me. I guess it has a stack trace now, but it might be nice to panic

Reply:
Yeah, we can panic here, maybe that's better.